### ISSUE TITLE
Crashes on some completions when no snippet engine is configured/installed

### ISSUE THREAD

**Issue by dccsillag** [hrsh7th/nvim-cmp#304]
I'll give an example -- with rust-analyzer, you start typing, and see the following completion:

```
print!(…)~       Method
```

Note the `~`. Upon triggering its completion, I get the following error:

```
E5108: Error executing lua ...te/pack/packer/start/nvim-cmp/lua/cmp/config/default.lua:27: snippet engine is not configured.
```

And, after that, complletions no longer work -- in particular, no autocomplete menu appears. I have to restart neovim to get them back.

Now, I don't like snippets, and so I really don't want to install any of that. While it might be a good idea to, when no snippet engine is available, not even show completions which require a snippet engine, I think a very reasonable solution would be to just show a friendly error that no snippet engine is available and keep going (that is, autocompletions should continue working after that single event).

**Comment by hrsh7th** [comment]
Sorry. I have no plan to support the usage without snippet engine.

**Comment by dmitmel** [comment]
Rust Analyzer in particular has a couple of options for disabling snippets in the completion results: `rust-analyzer.completion.addCallArgumentSnippets` and `rust-analyzer.completion.postfix.enable`. Set both to false, they should get rid of snippets altogether.

**Comment by dccsillag** [comment]
> Because some language server returns snippet even if the client doesn't advertises it.

I see.

> If you can define how to handle snippets without a snippet engine, we can support it.

I was thinking of something along the following line (disclaimer: I haven't looked at nvim-cmp's code yet):

On the "complete" operation (what `cmp.mapping.complete` represents), check if the server has returned a snippet or just text (i.e., not a snippet) (as I undestand it, this is already done). If it's just text, then follow the operations that are done today. If it's a snippet, then simply cancel the completion (`cmp.mapping.abort` or `cmp.mapping.close`, or by not doing the confirmation); in either case, it's probably a good idea to show a message (via `print`) to the user along the lines of "the confirmed completion requires snippets, but no snippet engine was configured".

As an outsider, this seems quite reasonable -- both as a user and implementation-wise. Thoughts?

**Comment by dmitmel** [comment]
@babygau Yes, if the server you are using doesn't respect client capabilities and doesn't have an option for turning off snippets in the completion results - yes.

**Comment by breuerfelix** [comment]
would be awesome if you could state in the readme, that it is highly recommended to use a snippet plugin or otherwhise some language servers will break.  
took me some time to get here :) i wanted a minimal solution so i thought i go without snippets, now i will add some
